### ISSUE TITLE
null pointer dereference in id3_ucs4_length

### ISSUE THREAD

**Issue by triffid** [tenacityteam/libid3tag#6]
https://access.redhat.com/security/cve/cve-2017-11550

https://bugs.gentoo.org/626698

I just ran into this today, and noticed that it seems to have been reported elsewhere but not here.

Seems to be a lack of pointer sanity checking in https://github.com/tenacityteam/libid3tag/blob/main/ucs4.c#L40-L43

Got this backtrace today during `mpd update`:

```
Thread 7 "update" received signal SIGSEGV, Segmentation fault.                                                                                                                                                                                               [Switching to Thread 0x7fffe6ffd640 (LWP 22570)]                                                                                                                                                                                                             0x00007ffff5d71a70 in id3_ucs4_length () from /usr/lib64/libid3tag.so.0.16.1                                                                                                                                                                                 (gdb) bt                                                                                                                                                                                                                                                     #0  0x00007ffff5d71a70 in id3_ucs4_length () at /usr/lib64/libid3tag.so.0.16.1                                                                                                                                                                               #1  0x00007ffff5d6cca6 in id3_compat_fixup () at /usr/lib64/libid3tag.so.0.16.1                                                                                                                                                                              #2  0x00007ffff5d7100c in id3_tag_parse () at /usr/lib64/libid3tag.so.0.16.1                                                                                                                                                                                 #3  0x0000555555609861 in  ()                                                                                                                                                                                                                                #4  0x0000555555609d4e in  ()                                                                                                                                                                                                                                #5  0x0000555555609da0 in  ()                                                                                                                                                                                                                                #6  0x000055555560a523 in  ()                                                                                                                                                                                                                                #7  0x00005555555b1381 in  ()                                                                                                                                                                                                                                #8  0x00005555555b1447 in  ()                                                                                                                                                                                                                                #9  0x00005555555af3aa in  ()                                                                                                                                                                                                                                #10 0x00005555555af4df in  ()                                                                                                                                                                                                                                #11 0x0000555555619b68 in  ()                                                                                                                                                                                                                                #12 0x0000555555619e60 in  ()                                                                                                                                                                                                                                #13 0x0000555555619474 in  ()                                                                                                                                                                                                                                #14 0x00005555556190ee in  ()                                                                                                                                                                                                                                #15 0x00005555556193e3 in  ()                                                                                                                                                                                                                                #16 0x00005555556190ee in  ()                                                                                                                                                                                                                                #17 0x00005555556193e3 in  ()                                                                                                                                                                                                                                #18 0x00005555556190ee in  ()                                                                                                                                                                                                                                #19 0x00005555556193e3 in  ()                                                                                                                                                                                                                                #20 0x00005555556190ee in  ()                                                                                                                                                                                                                                #21 0x00005555556198aa in  ()                                                                                                                                                                                                                                #22 0x0000555555616d86 in  ()                                                                                                                                                                                                                                #23 0x00005555555cb003 in  ()                                                                                                                                                                                                                                #24 0x00007ffff5383e3e in start_thread () at /lib64/libpthread.so.0                                                                                                                                                                                          #25 0x00007ffff52b92cf in clone () at /lib64/libc.so.6                                                                                                                                                                                                       (gdb) disassemble
Dump of assembler code for function id3_ucs4_length:
=> 0x00007ffff5d71a70 <+0>:     mov    (%rdi),%ecx
0x00007ffff5d71a72 <+2>:     test   %ecx,%ecx
0x00007ffff5d71a74 <+4>:     je     0x7ffff5d71a98 <id3_ucs4_length+40>
0x00007ffff5d71a76 <+6>:     mov    %rdi,%rax
0x00007ffff5d71a79 <+9>:     nopl   0x0(%rax)
0x00007ffff5d71a80 <+16>:    mov    0x4(%rax),%edx
0x00007ffff5d71a83 <+19>:    add    $0x4,%rax
0x00007ffff5d71a87 <+23>:    test   %edx,%edx
0x00007ffff5d71a89 <+25>:    jne    0x7ffff5d71a80 <id3_ucs4_length+16>
0x00007ffff5d71a8b <+27>:    sub    %rdi,%rax
0x00007ffff5d71a8e <+30>:    sar    $0x2,%rax
0x00007ffff5d71a92 <+34>:    ret    
0x00007ffff5d71a93 <+35>:    nopl   0x0(%rax,%rax,1)
0x00007ffff5d71a98 <+40>:    xor    %eax,%eax
0x00007ffff5d71a9a <+42>:    ret
End of assembler dump.
(gdb) i r
rax            0x0                 0
rbx            0x0                 0
rcx            0x7fffd020c5e1      140736685196769
rdx            0x0                 0
rsi            0x0                 0
rdi            0x0                 0
rbp            0x7fffd01e918c      0x7fffd01e918c
rsp            0x7fffe6fe9b68      0x7fffe6fe9b68
r8             0x0                 0
r9             0x7fffd01ef500      140736685077760
r10            0x7ffff5376ba0      140737307437984
r11            0x4                 4
r12            0x1                 1
r13            0x7fffd01eefc0      140736685076416
r14            0x7ffff5d74000      140737317912576
r15            0x7fffe6fe9b70      140737068833648
rip            0x7ffff5d71a70      0x7ffff5d71a70 <id3_ucs4_length>
eflags         0x10246             [ PF ZF IF RF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
```

The unknown symbols are functions within `mpd` fwiw, I haven't built it with debug symbols apparently.

I posted a test file at https://triffid-hunter.no-ip.info/test-id3taglib-segfault.mp3 in case it helps - although it seems like my music collection has numerous files that trigger this fault, and I have to use `strace` to find out which files are triggering it since `mpd` doesn't print the filename until _after_ id3taglib has looked at it.

Regenerating the file with `ffmpeg -c:a copy` seems to help fwiw, and both `ffmpeg` and `mpv` read/play the unaltered file without drama.

**Comment by Be-ing** [comment]
Thanks for reporting this. A pull request fixing this would be welcome.

**Comment by triffid** [comment]
I'd have to dig deeper into the codebase for that, work out 1) what this function should return if it finds a null, and/or 2) why it's stumbling across nulls in the first place - perhaps the actual bug is elsewhere, and the segfault only happens here symptomatically?

I might have a crack at it at some point, but ideally someone more familiar with the codebase could beat me to a solution.

**Comment by triffid** [comment]
Heh fair enough, well you may be interested to note that this repo is the id3taglib package upstream for at least Gentoo Linux, thanks for taking over stewardship!

I'll submit a pull at some point if no-one beats me to it ;)

**Comment by kleph** [comment]
Thanks for the patch, I just tested it, on gentoo also and it works well. I attached it to the issue in gentoo bug tracker.

Did not dig into the code to answer to know if it's just a consequence of another bug though

**Comment by triffid** [comment]
Oh thanks for reminding me about this @kleph!

@Be-ing several folks at https://bugs.gentoo.org/626698 seem to find that #7 solves this issue - is there any impediment to #7 being merged?

**Comment by Be-ing** [comment]
Fixed by #7 and [released in 0.16.2](https://github.com/tenacityteam/libid3tag/releases/tag/0.16.2)
